server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncProtector.java

```diff
@@ -0,0 +1,169 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.doublelive;
+
+import org.apache.iotdb.db.conf.IoTDBDescriptor;
+import org.apache.iotdb.db.engine.fileSystem.SystemFileFactory;
+import org.apache.iotdb.db.qp.physical.PhysicalPlan;
+import org.apache.iotdb.db.writelog.io.SingleFileLogReader;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public abstract class OperationSyncProtector implements Runnable {
+
+  protected static final Logger LOGGER = LoggerFactory.getLogger(OperationSyncProtector.class);
+  protected static final int logFileValidity =
+      IoTDBDescriptor.getInstance().getConfig().getOperationSyncLogValidity();
+
+  // For transmit log files
+  protected final Lock logFileListLock;
+  protected List<String> registeredLogFiles;
+  protected List<String> processingLogFiles;
+
+  // For serialize PhysicalPlan
+  private static final int MAX_PHYSICALPLAN_SIZE = 16 * 1024 * 1024;
+  protected final ByteArrayOutputStream protectorByteStream;
+  protected final DataOutputStream protectorDeserializeStream;
+
+  // Working state
+  protected volatile boolean isProtectorAtWork;
+
+  protected OperationSyncProtector() {
+    logFileListLock = new ReentrantLock();
+    registeredLogFiles = new ArrayList<>();
+
+    protectorByteStream = new ByteArrayOutputStream(MAX_PHYSICALPLAN_SIZE);
+    protectorDeserializeStream = new DataOutputStream(protectorByteStream);
+
+    isProtectorAtWork = false;
+  }
+
+  protected void registerLogFile(String logFile) {
+    logFileListLock.lock();
+    registeredLogFiles.add(logFile);
+    logFileListLock.unlock();
+  }
+
+  protected void wrapLogFiles() {
+    processingLogFiles = new ArrayList<>(registeredLogFiles);
+    registeredLogFiles = new ArrayList<>();
+  }
+
+  @Override
+  public void run() {
+    while (true) {
+      while (true) {
+        // Wrap and transmit all OperationSyncLogs
+        logFileListLock.lock();
+        if (registeredLogFiles.size() > 0) {
+          isProtectorAtWork = true;
+          wrapLogFiles();
+          logFileListLock.unlock();
```

Comment:
unlock should always be in finally block

Reply:
Fixed.